src/Microsoft.Diagnostics.Monitoring.StartupHook/Exceptions/Eventing/ExceptionsEventSource.cs

```diff
@@ -0,0 +1,215 @@
+            using PinnedData namePinned = PinnedData.Create(Name);
+            using PinnedData typeArgsPinned = PinnedData.Create(TypeArgs);
+
+            SetValue(ref data[NameIdentificationEvents.FunctionDescPayloads.FunctionId], FunctionId);
+            SetValue(ref data[NameIdentificationEvents.FunctionDescPayloads.ClassId], ClassId);
+            SetValue(ref data[NameIdentificationEvents.FunctionDescPayloads.ClassToken], ClassToken);
+            SetValue(ref data[NameIdentificationEvents.FunctionDescPayloads.ModuleId], ModuleId);
+            SetValue(ref data[NameIdentificationEvents.FunctionDescPayloads.Name], namePinned);
+            SetValue(ref data[NameIdentificationEvents.FunctionDescPayloads.TypeArgs], typeArgsPinned);
+
+            WriteEventCore(FunctionDescriptionEventId, data);
+        }
+
+        [Event(ModuleDescriptionEventId)]
+        public void WriteModuleDescription(
+            ulong ModuleId,
+            string Name)
+        {
+            Span<EventData> data = stackalloc EventData[2];
+            using PinnedData namePinned = PinnedData.Create(Name);
+
+            SetValue(ref data[NameIdentificationEvents.ModuleDescPayloads.ModuleId], ModuleId);
+            SetValue(ref data[NameIdentificationEvents.ModuleDescPayloads.Name], namePinned);
+
+            WriteEventCore(ModuleDescriptionEventId, data);
+        }
+
+        [Event(TokenDescriptionEventId)]
+        public void WriteTokenDescription(
+            ulong ModuleId,
+            uint Token,
+            uint OuterToken,
+            string Name)
+        {
+            Span<EventData> data = stackalloc EventData[4];
+            using PinnedData namePinned = PinnedData.Create(Name);
+
+            SetValue(ref data[NameIdentificationEvents.TokenDescPayloads.ModuleId], ModuleId);
+            SetValue(ref data[NameIdentificationEvents.TokenDescPayloads.Token], Token);
+            SetValue(ref data[NameIdentificationEvents.TokenDescPayloads.OuterToken], OuterToken);
+            SetValue(ref data[NameIdentificationEvents.TokenDescPayloads.Name], namePinned);
+
+            WriteEventCore(TokenDescriptionEventId, data);
+        }
+
+        [NonEvent]
+        private unsafe void WriteEventCore(int eventId, Span<EventData> data)
+        {
+            fixed (EventData* dataPtr = data)
+            {
+                WriteEventCore(eventId, data.Length, dataPtr);
+            }
+        }
+
+        [NonEvent]
+        private static unsafe void SetValue<T>(ref EventData data, in T value) where T : unmanaged
+        {
+            data.DataPointer = (nint)Unsafe.AsPointer(ref Unsafe.AsRef(value));
+            data.Size = sizeof(T);
+        }
+
+        [NonEvent]
+        private static void SetValue(ref EventData data, in PinnedData value)
+        {
+            data.DataPointer = value.Address;
+            data.Size = value.Size;
+        }
+
+        private ref struct PinnedData
```

Comment:
Should this implement IDisposable?

Reply:
Ref structs cannot implement interfaces. However, this doesn't _really_ need to be a ref struct; I think it might have had some pointers or refs in earlier iterations. I can change this to be a regular struct and have it implement `IDisposable`.

Reply:
Made this a non-ref struct and implemented IDisposable.